include/envoy/stream_info/filter_state.h

```diff
@@ -29,20 +32,42 @@ class FilterState {
   /**
    * @param data_name the name of the data being set.
    * @param data an owning pointer to the data to be stored.
-   * Note that it is an error to call setData() twice with the same data_name; this is to
-   * enforce a single authoritative source for each piece of data stored in FilterState.
+   * @param state_type indicates whether the object is mutable or not.
+   *
+   * Note that it is an error to call setData() twice with the same
+   * data_name, if the existing object is immutable. This is to enforce a
+   * single authoritative source for each piece of immutable data stored in
+   * FilterState.
    */
-  virtual void setData(absl::string_view data_name, std::unique_ptr<Object>&& data) PURE;
+  virtual void setData(absl::string_view data_name, std::unique_ptr<Object>&& data,
+                       StateType state_type) PURE;
```

Comment:
My personal preference would be to have two separate interfaces rather than passing in an (effective) boolean (I think it'd be harder to mis-use and easier to audit if a company wanted to forbid one of them) but I am not dead-set on the idea.

Reply:
I was hoping the enum would make things cleaner

Reply:
I think it's equally clear if we have `setDataImmutable` and `setDataMutable` vs. having the enum. So, I'm on the fence on this one. The `enum` does support more dynamic reflective behavior, but I don't think that's a strong selling point here. The getters don't use an enum... should we be consistent in naming and signatures?